runtime/ast/access.go

```diff
@@ -20,38 +20,156 @@ package ast
 
 import (
 	"encoding/json"
+	"strings"
 
 	"github.com/onflow/cadence/runtime/errors"
 )
 
-//go:generate go run golang.org/x/tools/cmd/stringer -type=Access
+//go:generate go run golang.org/x/tools/cmd/stringer -type=PrimitiveAccess
 
-type Access uint
+type Access interface {
+	isAccess()
+	Keyword() string
+	Description() string
+	String() string
+	MarshalJSON() ([]byte, error)
+	IsLessPermissiveThan(Access) bool
+}
+
+type EntitlementSet interface {
+	Entitlements() []*NominalType
+	Separator() string
+}
+
+type ConjunctiveEntitlementSet struct {
+	Elements []*NominalType `json:"ConjunctiveElements"`
+}
+
+func (s ConjunctiveEntitlementSet) Entitlements() []*NominalType {
+	return s.Elements
+}
+
+func (s ConjunctiveEntitlementSet) Separator() string {
+	return ","
+}
+
+func NewConjunctiveEntitlementSet(entitlements []*NominalType) ConjunctiveEntitlementSet {
+	return ConjunctiveEntitlementSet{Elements: entitlements}
+}
+
+type DisjunctiveEntitlementSet struct {
+	Elements []*NominalType `json:"DisjunctiveElements"`
+}
+
+func (s DisjunctiveEntitlementSet) Entitlements() []*NominalType {
+	return s.Elements
+}
+
+func (s DisjunctiveEntitlementSet) Separator() string {
+	return " |"
+}
+
+func NewDisjunctiveEntitlementSet(entitlements []*NominalType) DisjunctiveEntitlementSet {
+	return DisjunctiveEntitlementSet{Elements: entitlements}
+}
+
+type EntitlementAccess struct {
+	EntitlementSet EntitlementSet
+}
+
+var _ Access = EntitlementAccess{}
+
+func NewEntitlementAccess(entitlements EntitlementSet) EntitlementAccess {
+	return EntitlementAccess{EntitlementSet: entitlements}
+}
+
+func (EntitlementAccess) isAccess() {}
+
+func (EntitlementAccess) Description() string {
+	return "entitled access"
+}
+
+func (e EntitlementAccess) entitlementsString() string {
+	str := strings.Builder{}
+	for i, entitlement := range e.EntitlementSet.Entitlements() {
+		str.Write([]byte(entitlement.String()))
+		if i < len(e.EntitlementSet.Entitlements())-1 {
+			str.Write([]byte(e.EntitlementSet.Separator()))
+		}
+	}
+	return str.String()
+}
+
+func (e EntitlementAccess) String() string {
+	return "ConjunctiveEntitlementAccess " + e.entitlementsString()
+}
+
+func (e EntitlementAccess) Keyword() string {
+	return "access(" + e.entitlementsString() + ")"
+}
+
+func (e EntitlementAccess) MarshalJSON() ([]byte, error) {
+	return json.Marshal(e.String())
+}
+
+func (e EntitlementAccess) subset(other EntitlementAccess) bool {
+	otherSet := make(map[*NominalType]struct{})
+	for _, entitlement := range other.EntitlementSet.Entitlements() {
```

Comment:
Pre-allocate to the known final size of the map:

```suggestion
	otherEntitlements := other.EntitlementSet.Entitlements()
	otherSet := make(map[*NominalType]struct{}, len(otherEntitlements))
	for _, entitlement := range otherEntitlements {
```

Also, maybe the map should be cached in the `EntitlementAccess`? Maybe `ConjunctiveEntitlementSet.Elements` should be an `OrderedSet`?